File: mc_foc_sl_fip_dspic33ck_mclv2/mcc_generated_files/X2CCode/Library/Control/Conversion/C/src/ConvFnc_TF2_FiP16.c
/*
 * Copyright (c) 2013, Linz Center of Mechatronics GmbH (LCM) http://www.lcm.at/
 * All rights reserved.
 */
/*
 * This file is licensed according to the BSD 3-clause license as follows:
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the "Linz Center of Mechatronics GmbH" and "LCM" nor
 *       the names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL "Linz Center of Mechatronics GmbH" BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/*
 * This file is part of X2C. http://x2c.lcm.at/
 * $LastChangedRevision: 1975 $
 */
/* USERCODE-BEGIN:Description                                                                                         */
/* Description: */
/* USERCODE-END:Description                                                                                           */
#include "ConvFnc_TF2_FiP16.h"

/* USERCODE-BEGIN:PreProcessor                                                                                        */
/* USERCODE-END:PreProcessor                                                                                          */

extern uint8 MaskParamBuffer[];
extern uint8 ImplParamBuffer[];
extern uint8 SaveFncDataBuffer[];

/** Implementation parameter data structure */
typedef struct {
    int16 b0;
    int16 b1;
    int16 b2;
    int16 a0;
    int16 a1;
    int8 sfrb;
    int8 sfra;
} TF2_FIP16_IMPL_PARAM;

/** Private prototypes */
static uint8 convertM2I(const TF2_FIP16_MASK_PARAM *maskParam, TF2_FIP16_IMPL_PARAM *impParam, float_CoT ts);

/**
 * @brief Converts Mask parameters to Implementation parameters.
 *
 * @param[in] maskParam Mask parameters
 * @param[out] impParam Implementation parameters
 * @param[in] ts Block sample time
 *
 * @return Error: zero on success, not zero in case of conversion error
 */
static uint8 convertM2I(const TF2_FIP16_MASK_PARAM *maskParam, TF2_FIP16_IMPL_PARAM *impParam, float_CoT ts)
{
    uint8 error = (uint8)0;
/* USERCODE-BEGIN:Conversion                                                                                          */
    float_CoT mp_b0, mp_b1, mp_b2, mp_a0, mp_a1;
    uint8 sfrb, sfrb0, sfrb1, sfrb2, sfra, sfra0, sfra1;

    /* get controller parameter values */
    mp_b0 = (float_CoT)(maskParam->b0);
    mp_b1 = (float_CoT)(maskParam->b1);
    mp_b2 = (float_CoT)(maskParam->b2);
    mp_a0 = (float_CoT)(maskParam->a0);
    mp_a1 = (float_CoT)(maskParam->a1);


    /* calculate shift factors */
    sfrb0 = getQFormat(mp_b0, 16);
    sfrb1 = getQFormat(mp_b1, 16);
    sfrb2 = getQFormat(mp_b2, 16);
    sfrb = sfrb0;
    if (sfrb > sfrb1)
    {
        sfrb = sfrb1;
    }
    if (sfrb > sfrb2)
    {
        sfrb = sfrb2;
    }
    sfra0 = getQFormat(mp_a0, 16);
    sfra1 = getQFormat(mp_a1, 16);
    if (sfra0 > sfra1)
    {
        sfra = sfra1;
    }
    else
    {
        sfra = sfra0;
    }

    /* set shift factors and Q values */
    impParam->sfrb = sfrb;
    impParam->sfra = sfra;
    impParam->b0 = getQx16Value(mp_b0, sfrb);
    impParam->b1 = getQx16Value(mp_b1, sfrb);
    impParam->b2 = getQx16Value(mp_b2, sfrb);
    impParam->a0 = getQx16Value(mp_a0, sfrb);
    impParam->a1 = getQx16Value(mp_a1, sfrb);

/* USERCODE-END:Conversion                                                                                            */
    return (error);
}

/**
 * @brief Load block mask parameter data.
 *
 * @param[in] maskParam Mask parameter data structure
 * @param[out] data Data
 * @param[out] dataLen Data length
 * @param[in] maxSize Maximum Service data length
 *
 * @return Error: zero on success, not zero in case of load- or conversion error
 */
uint8 TF2_FiP16_LoadMP(const TF2_FIP16_MASK_PARAM *maskParam, uint8 data[], uint16* dataLen, uint16 maxSize)
{
    uint8 error = (uint8)0;
    if ((uint16)48 > maxSize)
    {
        error = (uint8)1;
    }
    else
    {
        data[0] = (uint8)((*(uint64*)&(maskParam->b2)) & 0x00000000000000FF);
        data[1] = (uint8)((*(uint64*)&(maskParam->b2) >> 8) & 0x00000000000000FF);
        data[2] = (uint8)((*(uint64*)&(maskParam->b2) >> 16) & 0x00000000000000FF);
        data[3] = (uint8)((*(uint64*)&(maskParam->b2) >> 24) & 0x00000000000000FF);
        data[4] = (uint8)((*(uint64*)&(maskParam->b2) >> 32) & 0x00000000000000FF);
        data[5] = (uint8)((*(uint64*)&(maskParam->b2) >> 40) & 0x00000000000000FF);
        data[6] = (uint8)((*(uint64*)&(maskParam->b2) >> 48) & 0x00000000000000FF);
        data[7] = (uint8)((*(uint64*)&(maskParam->b2) >> 56) & 0x00000000000000FF);
        data[8] = (uint8)((*(uint64*)&(maskParam->b1)) & 0x00000000000000FF);
        data[9] = (uint8)((*(uint64*)&(maskParam->b1) >> 8) & 0x00000000000000FF);
        data[10] = (uint8)((*(uint64*)&(maskParam->b1) >> 16) & 0x00000000000000FF);
        data[11] = (uint8)((*(uint64*)&(maskParam->b1) >> 24) & 0x00000000000000FF);
        data[12] = (uint8)((*(uint64*)&(maskParam->b1) >> 32) & 0x00000000000000FF);
        data[13] = (uint8)((*(uint64*)&(maskParam->b1) >> 40) & 0x00000000000000FF);
        data[14] = (uint8)((*(uint64*)&(maskParam->b1) >> 48) & 0x00000000000000FF);
        data[15] = (uint8)((*(uint64*)&(maskParam->b1) >> 56) & 0x00000000000000FF);
        data[16] = (uint8)((*(uint64*)&(maskParam->b0)) & 0x00000000000000FF);
        data[17] = (uint8)((*(uint64*)&(maskParam->b0) >> 8) & 0x00000000000000FF);
        data[18] = (uint8)((*(uint64*)&(maskParam->b0) >> 16) & 0x00000000000000FF);
        data[19] = (uint8)((*(uint64*)&(maskParam->b0) >> 24) & 0x00000000000000FF);
        data[20] = (uint8)((*(uint64*)&(maskParam->b0) >> 32) & 0x00000000000000FF);
        data[21] = (uint8)((*(uint64*)&(maskParam->b0) >> 40) & 0x00000000000000FF);
        data[22] = (uint8)((*(uint64*)&(maskParam->b0) >> 48) & 0x00000000000000FF);
        data[23] = (uint8)((*(uint64*)&(maskParam->b0) >> 56) & 0x00000000000000FF);
        data[24] = (uint8)((*(uint64*)&(maskParam->a1)) & 0x00000000000000FF);
        data[25] = (uint8)((*(uint64*)&(maskParam->a1) >> 8) & 0x00000000000000FF);
        data[26] = (uint8)((*(uint64*)&(maskParam->a1) >> 16) & 0x00000000000000FF);
        data[27] = (uint8)((*(uint64*)&(maskParam->a1) >> 24) & 0x00000000000000FF);
        data[28] = (uint8)((*(uint64*)&(maskParam->a1) >> 32) & 0x00000000000000FF);
        data[29] = (uint8)((*(uint64*)&(maskParam->a1) >> 40) & 0x00000000000000FF);
        data[30] = (uint8)((*(uint64*)&(maskParam->a1) >> 48) & 0x00000000000000FF);
        data[31] = (uint8)((*(uint64*)&(maskParam->a1) >> 56) & 0x00000000000000FF);
        data[32] = (uint8)((*(uint64*)&(maskParam->a0)) & 0x00000000000000FF);
        data[33] = (uint8)((*(uint64*)&(maskParam->a0) >> 8) & 0x00000000000000FF);
        data[34] = (uint8)((*(uint64*)&(maskParam->a0) >> 16) & 0x00000000000000FF);
        data[35] = (uint8)((*(uint64*)&(maskParam->a0) >> 24) & 0x00000000000000FF);
        data[36] = (uint8)((*(uint64*)&(maskParam->a0) >> 32) & 0x00000000000000FF);
        data[37] = (uint8)((*(uint64*)&(maskParam->a0) >> 40) & 0x00000000000000FF);
        data[38] = (uint8)((*(uint64*)&(maskParam->a0) >> 48) & 0x00000000000000FF);
        data[39] = (uint8)((*(uint64*)&(maskParam->a0) >> 56) & 0x00000000000000FF);
        data[40] = (uint8)((*(uint64*)&(maskParam->ts_fact)) & 0x00000000000000FF);
        data[41] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 8) & 0x00000000000000FF);
        data[42] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 16) & 0x00000000000000FF);
        data[43] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 24) & 0x00000000000000FF);
        data[44] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 32) & 0x00000000000000FF);
        data[45] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 40) & 0x00000000000000FF);
        data[46] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 48) & 0x00000000000000FF);
        data[47] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 56) & 0x00000000000000FF);
        *dataLen = (uint16)48;
    }
    return (error);
}

/**
 * @brief Save block mask parameter data.
 *
 * @param[in] block Pointer to block structure
 * @param[out] maskParam Mask parameter data structure
 * @param[in] data Data
 * @param dataLen Length of mask parameter data stream
 *
 * @return Error: zero on success, not zero in case of save- or conversion error
 */
uint8 TF2_FiP16_SaveMP(TF2_FIP16 *block, TF2_FIP16_MASK_PARAM *maskParam, const uint8 data[], uint16 dataLen)
{
    uint8 error = (uint8)0;

    if (dataLen != (uint16)48)
    {
        error = (uint8)1;
    }
    else
    {
        uint64 tmp64;
        /** - Cache old mask parameters */
        TF2_FiP16_BackupMP(maskParam);

        /** - Save new mask parameters */
        tmp64 = (uint64)data[0] + \
            ((uint64)data[1] << 8) + ((uint64)data[2] << 16) + \
            ((uint64)data[3] << 24) + ((uint64)data[4] << 32) + \
            ((uint64)data[5] << 40) + ((uint64)data[6] << 48) + \
            ((uint64)data[7] << 56);
        maskParam->b2 = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[8] + \
            ((uint64)data[9] << 8) + ((uint64)data[10] << 16) + \
            ((uint64)data[11] << 24) + ((uint64)data[12] << 32) + \
            ((uint64)data[13] << 40) + ((uint64)data[14] << 48) + \
            ((uint64)data[15] << 56);
        maskParam->b1 = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[16] + \
            ((uint64)data[17] << 8) + ((uint64)data[18] << 16) + \
            ((uint64)data[19] << 24) + ((uint64)data[20] << 32) + \
            ((uint64)data[21] << 40) + ((uint64)data[22] << 48) + \
            ((uint64)data[23] << 56);
        maskParam->b0 = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[24] + \
            ((uint64)data[25] << 8) + ((uint64)data[26] << 16) + \
            ((uint64)data[27] << 24) + ((uint64)data[28] << 32) + \
            ((uint64)data[29] << 40) + ((uint64)data[30] << 48) + \
            ((uint64)data[31] << 56);
        maskParam->a1 = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[32] + \
            ((uint64)data[33] << 8) + ((uint64)data[34] << 16) + \
            ((uint64)data[35] << 24) + ((uint64)data[36] << 32) + \
            ((uint64)data[37] << 40) + ((uint64)data[38] << 48) + \
            ((uint64)data[39] << 56);
        maskParam->a0 = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[40] + \
            ((uint64)data[41] << 8) + ((uint64)data[42] << 16) + \
            ((uint64)data[43] << 24) + ((uint64)data[44] << 32) + \
            ((uint64)data[45] << 40) + ((uint64)data[46] << 48) + \
            ((uint64)data[47] << 56);
        maskParam->ts_fact = (float64)(*(float64*)&tmp64);

        error = TF2_FiP16_ConvertMP(block, maskParam);

        /** - Check for errors during execution of save function */
        if (error != 0)
        {
            /* Restore old mask parameter */
            TF2_FiP16_RestoreMP(maskParam);
        }

    }
    return (error);
}

/**
 * @brief Initializes Implementation parameters from Mask parameters.
 *
 * @param[out] block Block
 * @param[in] maskParam Mask parameters
 *
 * @return Error: zero on success, not zero in case of initialization error
 */
uint8 TF2_FiP16_InitMP(TF2_FIP16 *block, const TF2_FIP16_MASK_PARAM *maskParam)
{
    uint8 error = (uint8)0;
    float_CoT block_ts;

    TF2_FIP16_IMPL_PARAM *implParam = (TF2_FIP16_IMPL_PARAM*)ImplParamBuffer;

    /** - Get Block Sample Time */
    block_ts = maskParam->ts_fact * getModelSampleTime();
    /** - Convert mask parameters to implementation parameters */
    error = convertM2I(maskParam, implParam, block_ts);

    /** - Prepare implementation parameters for save function */
    SaveFncDataBuffer[0] = (uint8)(implParam->b0 & 0x00FF);
    SaveFncDataBuffer[1] = (uint8)((implParam->b0 >> 8) & 0x00FF);
    SaveFncDataBuffer[2] = (uint8)(implParam->b1 & 0x00FF);
    SaveFncDataBuffer[3] = (uint8)((implParam->b1 >> 8) & 0x00FF);
    SaveFncDataBuffer[4] = (uint8)(implParam->b2 & 0x00FF);
    SaveFncDataBuffer[5] = (uint8)((implParam->b2 >> 8) & 0x00FF);
    SaveFncDataBuffer[6] = (uint8)(implParam->a0 & 0x00FF);
    SaveFncDataBuffer[7] = (uint8)((implParam->a0 >> 8) & 0x00FF);
    SaveFncDataBuffer[8] = (uint8)(implParam->a1 & 0x00FF);
    SaveFncDataBuffer[9] = (uint8)((implParam->a1 >> 8) & 0x00FF);
    SaveFncDataBuffer[10] = (uint8)implParam->sfrb;
    SaveFncDataBuffer[11] = (uint8)implParam->sfra;

    /** - Execute save function of block */
    error = TF2_FiP16_Save(block, SaveFncDataBuffer, 12);

    return (error);
}

/**
 * @brief Converts Mask parameters into Implementation parameters and executes Block Save function.
 *
 * @param[out] block Block Implementation
 * @param[in] maskParam Mask Parameters
 *
 * @return 0 if successful, greater than 0 if conversion failed
 */
uint8 TF2_FiP16_ConvertMP(TF2_FIP16 *block, TF2_FIP16_MASK_PARAM *maskParam)
{
    uint8 error;
    float_CoT block_ts;
    TF2_FIP16_IMPL_PARAM *implParam = (TF2_FIP16_IMPL_PARAM*)ImplParamBuffer;

    /** - Get Block Sample Time */
    block_ts = maskParam->ts_fact * getModelSampleTime();
    /** - Convert mask parameters to implementation parameters */
    error = convertM2I(maskParam, implParam, block_ts);
    if (error)
    {
        return (error);
    }

    /** - Prepare implementation parameters for save function */
    SaveFncDataBuffer[0] = (uint8)(implParam->b0 & 0x00FF);
    SaveFncDataBuffer[1] = (uint8)((implParam->b0 >> 8) & 0x00FF);
    SaveFncDataBuffer[2] = (uint8)(implParam->b1 & 0x00FF);
    SaveFncDataBuffer[3] = (uint8)((implParam->b1 >> 8) & 0x00FF);
    SaveFncDataBuffer[4] = (uint8)(implParam->b2 & 0x00FF);
    SaveFncDataBuffer[5] = (uint8)((implParam->b2 >> 8) & 0x00FF);
    SaveFncDataBuffer[6] = (uint8)(implParam->a0 & 0x00FF);
    SaveFncDataBuffer[7] = (uint8)((implParam->a0 >> 8) & 0x00FF);
    SaveFncDataBuffer[8] = (uint8)(implParam->a1 & 0x00FF);
    SaveFncDataBuffer[9] = (uint8)((implParam->a1 >> 8) & 0x00FF);
    SaveFncDataBuffer[10] = (uint8)implParam->sfrb;
    SaveFncDataBuffer[11] = (uint8)implParam->sfra;

    /** - Execute save function of block */
    error = TF2_FiP16_Save(block, SaveFncDataBuffer, 12);

    return (error);
}

/**
 * @brief Backups current Mask parameters into global Mask parameter backup array.
 *
 * @param[in] maskParam Mask parameters
 *
 * @return Nothing
 */
void TF2_FiP16_BackupMP(TF2_FIP16_MASK_PARAM* maskParam)
{
    memcpy(MaskParamBuffer, maskParam, sizeof(*maskParam));
}

/**
 * @brief Restores Mask parameters from global Mask parameter backup array.
 *
 * @param[out] maskParam Mask parameters
 *
 * @return Nothing
 */
void TF2_FiP16_RestoreMP(TF2_FIP16_MASK_PARAM* maskParam)
{
    memcpy(maskParam, MaskParamBuffer, sizeof(*maskParam));
}

File: mc_foc_sl_fip_dspic33ck_mclv2/mcc_generated_files/X2CCode/Library/Control/Conversion/Java/src/at/lcm/x2c/library/control/ConvFnc_TF2_FiP16.java
/*
 * Copyright (c) 2013, Linz Center of Mechatronics GmbH (LCM) http://www.lcm.at/
 * All rights reserved.
 */
/*
 * This file is licensed according to the BSD 3-clause license as follows:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the "Linz Center of Mechatronics GmbH" and "LCM" nor
 *       the names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL "Linz Center of Mechatronics GmbH" BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/*
 * This file is part of X2C. http://x2c.lcm.at/
 * $LastChangedRevision: 2568 $
 */
// USERCODE-BEGIN:Description
//*     Description:                                                           *
// USERCODE-END:Description

package at.lcm.x2c.library.control;

import at.lcm.x2c.core.structure.*;
import at.lcm.bu21.general.dtypes.*;

// USERCODE-BEGIN:Imports
import at.lcm.x2c.utils.QFormat;

// USERCODE-END:Imports

@SuppressWarnings("unused")
public class ConvFnc_TF2_FiP16 extends JavaConversionFunction {
    private static final long serialVersionUID = 1L;

    @Override
    public void convertMaskToImplementation() throws Exception {

        // Mask Parameter: b2
        MaskDouble b2MaskVal =
                (MaskDouble)this.getMaskParameter("b2").getMaskDataType();
        TNumeric b2MaskData =
                (TNumeric)b2MaskVal.getData();
        // Mask Parameter: b1
        MaskDouble b1MaskVal =
                (MaskDouble)this.getMaskParameter("b1").getMaskDataType();
        TNumeric b1MaskData =
                (TNumeric)b1MaskVal.getData();
        // Mask Parameter: b0
        MaskDouble b0MaskVal =
                (MaskDouble)this.getMaskParameter("b0").getMaskDataType();
        TNumeric b0MaskData =
                (TNumeric)b0MaskVal.getData();
        // Mask Parameter: a1
        MaskDouble a1MaskVal =
                (MaskDouble)this.getMaskParameter("a1").getMaskDataType();
        TNumeric a1MaskData =
                (TNumeric)a1MaskVal.getData();
        // Mask Parameter: a0
        MaskDouble a0MaskVal =
                (MaskDouble)this.getMaskParameter("a0").getMaskDataType();
        TNumeric a0MaskData =
                (TNumeric)a0MaskVal.getData();
        // Mask Parameter: ts_fact
        MaskDouble ts_factMaskVal =
                (MaskDouble)this.getMaskParameter("ts_fact").getMaskDataType();
        TNumeric ts_factMaskData =
                (TNumeric)ts_factMaskVal.getData();

        // Implementation Parameter: b0
        TNumeric b0CtrVal = this.getImplementationParameterData("b0");
        // Implementation Parameter: b1
        TNumeric b1CtrVal = this.getImplementationParameterData("b1");
        // Implementation Parameter: b2
        TNumeric b2CtrVal = this.getImplementationParameterData("b2");
        // Implementation Parameter: a0
        TNumeric a0CtrVal = this.getImplementationParameterData("a0");
        // Implementation Parameter: a1
        TNumeric a1CtrVal = this.getImplementationParameterData("a1");
        // Implementation Parameter: sfrb
        TNumeric sfrbCtrVal = this.getImplementationParameterData("sfrb");
        // Implementation Parameter: sfra
        TNumeric sfraCtrVal = this.getImplementationParameterData("sfra");
        // Implementation Parameter: in_old
        TNumeric in_oldCtrVal = this.getImplementationParameterData("in_old");
        // Implementation Parameter: in_veryold
        TNumeric in_veryoldCtrVal = this.getImplementationParameterData("in_veryold");
        // Implementation Parameter: out_old
        TNumeric out_oldCtrVal = this.getImplementationParameterData("out_old");
        // Implementation Parameter: out_veryold
        TNumeric out_veryoldCtrVal = this.getImplementationParameterData("out_veryold");

        // USERCODE-BEGIN:ConvMaskToImplementation
        double b0, b1, b2, a0, a1;
        int sfrb, sfrb0, sfrb1, sfrb2, sfra, sfra0, sfra1;
        final int BITS = 16;

        // get controller parameter values
        b0 = Double.valueOf(b0MaskVal.getValue());
        b1 = Double.valueOf(b1MaskVal.getValue());
        b2 = Double.valueOf(b2MaskVal.getValue());
        a0 = Double.valueOf(a0MaskVal.getValue());
        a1 = Double.valueOf(a1MaskVal.getValue());

        // validate controller parameters

        // calculate shift factor
        sfrb0 = QFormat.getQFormat(b0, BITS, true);
        sfrb1 = QFormat.getQFormat(b1, BITS, true);
        sfrb2 = QFormat.getQFormat(b2, BITS, true);
        sfrb = Math.min(sfrb0, Math.min(sfrb1, sfrb2));

        sfra0 = QFormat.getQFormat(a0, BITS, true);
        sfra1 = QFormat.getQFormat(a1, BITS, true);
        sfra = Math.min(sfra0, sfra1);

        // set shift factors and Q values
        sfrbCtrVal.setReal(0, 0, sfrb);
        sfraCtrVal.setReal(0, 0, sfra);
        b0CtrVal.setReal(0, 0, Double.valueOf(QFormat.getQValue(b0, (int) (sfrbCtrVal.getReal(0, 0)), BITS, true)));
        b1CtrVal.setReal(0, 0, Double.valueOf(QFormat.getQValue(b1, (int) (sfrbCtrVal.getReal(0, 0)), BITS, true)));
        b2CtrVal.setReal(0, 0, Double.valueOf(QFormat.getQValue(b2, (int) (sfrbCtrVal.getReal(0, 0)), BITS, true)));
        a0CtrVal.setReal(0, 0, Double.valueOf(QFormat.getQValue(a0, (int) (sfraCtrVal.getReal(0, 0)), BITS, true)));
        a1CtrVal.setReal(0, 0, Double.valueOf(QFormat.getQValue(a1, (int) (sfraCtrVal.getReal(0, 0)), BITS, true)));

        // USERCODE-END:ConvMaskToImplementation
        this.setImplementationParameterData("b0", b0CtrVal);
        this.setImplementationParameterData("b1", b1CtrVal);
        this.setImplementationParameterData("b2", b2CtrVal);
        this.setImplementationParameterData("a0", a0CtrVal);
        this.setImplementationParameterData("a1", a1CtrVal);
        this.setImplementationParameterData("sfrb", sfrbCtrVal);
        this.setImplementationParameterData("sfra", sfraCtrVal);
        this.setImplementationParameterData("in_old", in_oldCtrVal);
        this.setImplementationParameterData("in_veryold", in_veryoldCtrVal);
        this.setImplementationParameterData("out_old", out_oldCtrVal);
        this.setImplementationParameterData("out_veryold", out_veryoldCtrVal);
    }

    @Override
    public void convertImplementationToMask() throws Exception {

        // Mask Parameter: b2
        MaskDouble b2MaskVal =
                (MaskDouble)this.getMaskParameter("b2").getMaskDataType();
        TNumeric b2MaskData =
                (TNumeric)b2MaskVal.getData();
        // Mask Parameter: b1
        MaskDouble b1MaskVal =
                (MaskDouble)this.getMaskParameter("b1").getMaskDataType();
        TNumeric b1MaskData =
                (TNumeric)b1MaskVal.getData();
        // Mask Parameter: b0
        MaskDouble b0MaskVal =
                (MaskDouble)this.getMaskParameter("b0").getMaskDataType();
        TNumeric b0MaskData =
                (TNumeric)b0MaskVal.getData();
        // Mask Parameter: a1
        MaskDouble a1MaskVal =
                (MaskDouble)this.getMaskParameter("a1").getMaskDataType();
        TNumeric a1MaskData =
                (TNumeric)a1MaskVal.getData();
        // Mask Parameter: a0
        MaskDouble a0MaskVal =
                (MaskDouble)this.getMaskParameter("a0").getMaskDataType();
        TNumeric a0MaskData =
                (TNumeric)a0MaskVal.getData();
        // Mask Parameter: ts_fact
        MaskDouble ts_factMaskVal =
                (MaskDouble)this.getMaskParameter("ts_fact").getMaskDataType();
        TNumeric ts_factMaskData =
                (TNumeric)ts_factMaskVal.getData();

        // Implementation Parameter: b0
        TNumeric b0CtrVal = this.getImplementationParameterData("b0");
        // Implementation Parameter: b1
        TNumeric b1CtrVal = this.getImplementationParameterData("b1");
        // Implementation Parameter: b2
        TNumeric b2CtrVal = this.getImplementationParameterData("b2");
        // Implementation Parameter: a0
        TNumeric a0CtrVal = this.getImplementationParameterData("a0");
        // Implementation Parameter: a1
        TNumeric a1CtrVal = this.getImplementationParameterData("a1");
        // Implementation Parameter: sfrb
        TNumeric sfrbCtrVal = this.getImplementationParameterData("sfrb");
        // Implementation Parameter: sfra
        TNumeric sfraCtrVal = this.getImplementationParameterData("sfra");
        // Implementation Parameter: in_old
        TNumeric in_oldCtrVal = this.getImplementationParameterData("in_old");
        // Implementation Parameter: in_veryold
        TNumeric in_veryoldCtrVal = this.getImplementationParameterData("in_veryold");
        // Implementation Parameter: out_old
        TNumeric out_oldCtrVal = this.getImplementationParameterData("out_old");
        // Implementation Parameter: out_veryold
        TNumeric out_veryoldCtrVal = this.getImplementationParameterData("out_veryold");

        // USERCODE-BEGIN:ConvImplementationToMask
		double b0, b1, b2, a0, a1, sfrb, sfra;
		final int BITS = 16;

		// get controller parameter
		b0 = b0CtrVal.getReal(0, 0);
		b1 = b1CtrVal.getReal(0, 0);
		b2 = b2CtrVal.getReal(0, 0);
		a0 = a0CtrVal.getReal(0, 0);
		a1 = a1CtrVal.getReal(0, 0);
		sfrb = sfrbCtrVal.getReal(0, 0);
		sfra = sfraCtrVal.getReal(0, 0);

		// calculate controller coefficients
		b0 = QFormat.getDecValue((long) b0, (int) sfrb, BITS, true);
		b1 = QFormat.getDecValue((long) b1, (int) sfrb, BITS, true);
		b2 = QFormat.getDecValue((long) b2, (int) sfrb, BITS, true);
		a0 = QFormat.getDecValue((long) a0, (int) sfra, BITS, true);
		a1 = QFormat.getDecValue((long) a1, (int) sfra, BITS, true);

		// set mask parameter
		b0MaskData.setReal(0, 0, b0);
		b1MaskData.setReal(0, 0, b1);
		b2MaskData.setReal(0, 0, b2);
		a0MaskData.setReal(0, 0, a0);
		a1MaskData.setReal(0, 0, a1);

        // USERCODE-END:ConvImplementationToMask
        this.getMaskParameter("b2").setValue(b2MaskData.toFullString());
        this.getMaskParameter("b1").setValue(b1MaskData.toFullString());
        this.getMaskParameter("b0").setValue(b0MaskData.toFullString());
        this.getMaskParameter("a1").setValue(a1MaskData.toFullString());
        this.getMaskParameter("a0").setValue(a0MaskData.toFullString());
        this.getMaskParameter("ts_fact").setValue(ts_factMaskData.toFullString());
    }
}
